mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -511,19 +510,59 @@ private Optional<Map<TaskExecutorID, TaskExecutorState>> findTaskExecutorsFor(Ta
      * Finds the best fit Task Executor Group Key from the current set of executors
      * based on the provided `requestedConstraints`.
      *
-     * It filters the keys such that they satisfy the `requestedConstraints` and have a positive fitness score,
-     * then decides the best fit based on the highest fitness score.
+     * The method filters the groups that satisfy the `requestedConstraints`,
+     * then decides the best fit based on either matching `sizeName`
+     * or highest fitness score.
      *
-     * @param requestedConstraints Constraints of the scheduling request, serving as the reference for determining the best fit.
+     * @param requestedConstraints Constraints of the scheduling request,
+     * serving as the reference for determining the best fit.
      *
-     * @return An Optional wrapping the Task Executor Group Key with the best fit constraints. If no fitting constraints
-     * are found, it returns an empty Optional.
+     * @return An Optional wrapping the Task Executor Group Key with the best fit constraints.
+     * If no fitting constraints are found, it returns an empty Optional.
      */
     private Optional<TaskExecutorGroupKey> findBestGroup(SchedulingConstraints requestedConstraints) {
-        return executorsByGroup.keySet()
+        // Partition the keys into two sets: those that match the sizeName and those that don't
```

Comment:
suggestion: simplify this logic without storing an extra intermediate map of partition results. e.g.:
1. check for exact match on sizeName if present and return if found.
2. if not found scan with fitness score.

Reply:
I agree. I polished this function a bit. I hope it's more clear now.